### PR TITLE
Eip6110 queue deposit requests

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

 Feature

**What does this PR do? Why is it needed?**

implements https://github.com/ethereum/consensus-specs/pull/3818

- updates spectests to v1.5.0-alpha.8

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have made an appropriate entry to [CHANGELOG.md](https://github.com/prysmaticlabs/prysm/blob/develop/CHANGELOG.md).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

### CODE REVIEW

File: beacon-chain/core/electra/deposits.go [hunk 1]
```diff
@@ -238,68 +268,136 @@ func ProcessPendingBalanceDeposits(ctx context.Context, st state.BeaconState, ac
...
+//	else:
+//	    validator_index = ValidatorIndex(validator_pubkeys.index(deposit.pubkey))
+//	    # Increase balance
+//	    increase_balance(state, validator_index, deposit.amount)
+func ApplyPendingDeposit(ctx context.Context, st state.BeaconState, deposit *ethpb.PendingDeposit) error {
+	_, span := trace.StartSpan(ctx, "electra.ApplyPendingDeposit")
+	defer span.End()
+	index, ok := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(deposit.PublicKey))
+	if !ok {
+		valid, err := blocks.IsValidDepositSignature(&ethpb.Deposit_Data{
```
Comment (james-prysm) [inline beacon-chain/core/electra/deposits.go:384]: I wonder if i should be doing any batch verification before hand ...

File: beacon-chain/core/electra/validator.go [hunk 2]
```diff
@@ -102,30 +38,45 @@ func SwitchToCompoundingValidator(s state.BeaconState, idx primitives.ValidatorI
...
+//		        excess_balance = balance - MIN_ACTIVATION_BALANCE
+//		        state.balances[index] = MIN_ACTIVATION_BALANCE
+//		        validator = state.validators[index]
+//	         state.pending_deposits.append(PendingDeposit(
+//	           pubkey=validator.pubkey,
+//	           withdrawal_credentials=validator.withdrawal_credentials,
+//	           amount=excess_balance,
+//	           signature=bls.G2_POINT_AT_INFINITY,
+//	           slot=GENESIS_SLOT,
+//	         ))
```
Comment (terencechain) [inline beacon-chain/core/electra/validator.go:57]: the format here looks a bit shifted to the left by a few spaces

Comment (terencechain) [inline beacon-chain/core/electra/validator.go:57]: the format here looks a bit shifted to the left by a few spaces

Comment (terencechain) [reply beacon-chain/core/electra/validator.go:57]: same with other places

**Review by terencechain — COMMENTED** [review]
Thanks taking this one! I believe that using `ApplyPendingDeposit` in this case might introduce additional complexity and make the code a bit harder to follow, especially when passing `true`. We could streamline the process by implementing the following approach:
1. `AddValidatorToRegistry`
2. `IncreaseBalance`

These two functions can then be called directly from `ProcessPendingDeposits`, reducing the need for three separate instances of `ApplyPendingDeposit` within the function.

File: WORKSPACE [hunk 3]
```diff
@@ -290,7 +290,7 @@ filegroup(
     visibility = ["//visibility:public"],
 )
     """,
-    integrity = "sha256-BoXckDxXnDcEmAjg/dQgf/tLiJsb6CT0aZvmWHFijrY=",
+    integrity = "sha256-D/HPAW61lKqjoWwl7N0XvhdX+67dCEFAy8JxVzqBGtU=",
```
Comment (james-prysm) [inline WORKSPACE:293]: can someone double check this one for me

File: beacon-chain/core/electra/deposits.go [hunk 4]
```diff
@@ -238,99 +269,271 @@ func ProcessPendingBalanceDeposits(ctx context.Context, st state.BeaconState, ac
...
+				return errors.Wrap(err, "could not get validator")
+			}
+			isValidatorExited = val.ExitEpoch() < ffe
+			isValidatorWithdrawn = val.WithdrawableEpoch() < nextEpoch
+		}
+
+		if isValidatorWithdrawn {
+			// Deposited balance will never become active. Increase balance but do not consume churn
+			if err := ApplyPendingDeposit(ctx, st, pendingDeposit); err != nil {
+				return errors.Wrap(err, "could not apply pending deposit")
```
Comment (terencechain) [inline beacon-chain/core/electra/deposits.go:326]: I think you can just increase balance directly here

File: beacon-chain/core/electra/deposits.go [hunk 5]
```diff
@@ -238,99 +269,271 @@ func ProcessPendingBalanceDeposits(ctx context.Context, st state.BeaconState, ac
...
+	return st.SetDepositBalanceToConsume(0)
+}
+
+// batchProcessNewPendingDeposits should only be used to process new deposits that require validator registration
+func batchProcessNewPendingDeposits(ctx context.Context, st state.BeaconState, depositsToVerify []*ethpb.PendingDeposit) error {
+	if len(depositsToVerify) > 0 {
+		verified, err := blocks.BatchVerifyPendingDepositsSignatures(ctx, depositsToVerify)
+		if err != nil {
+			return errors.Wrap(err, "could not batch verify deposit signatures")
+		}
```
Comment (terencechain) [inline beacon-chain/core/electra/deposits.go:382]: if it's `verified`, you could just return early?

File: beacon-chain/core/electra/deposits.go [hunk 6]
```diff
@@ -238,99 +269,271 @@ func ProcessPendingBalanceDeposits(ctx context.Context, st state.BeaconState, ac
...
+	}
+	// Accumulate churn only if the churn limit has been hit.
+	if isChurnLimitReached {
+		return st.SetDepositBalanceToConsume(availableForProcessing - primitives.Gwei(processedAmount))
+	}
+	return st.SetDepositBalanceToConsume(0)
+}
+
+// batchProcessNewPendingDeposits should only be used to process new deposits that require validator registration
+func batchProcessNewPendingDeposits(ctx context.Context, st state.BeaconState, depositsToVerify []*ethpb.PendingDeposit) error {
```
Comment (terencechain) [inline beacon-chain/core/electra/deposits.go:377]: Tried making this more readable, feel free to use it

```go
func batchProcessNewPendingDeposits(ctx context.Context, state state.BeaconState, deposits []*ethpb.PendingDeposit) error {
	// Return early if there are no deposits to process
	if len(deposits) == 0 {
		return nil
	}

	// Try batch verification of all deposit signatures
	allSignaturesVerified, err := blocks.BatchVerifyPendingDepositsSignatures(ctx, deposits)
	if err != nil {
		return errors.Wrap(err, "batch signature verification failed")
	}

	// Process each deposit individually
	for _, deposit := range deposits {
		validSignature := allSignaturesVerified

		// If batch verification failed, check the individual deposit signature
		if !allSignaturesVerified {
			validSignature, err = blocks.IsValidDepositSignature(&ethpb.Deposit_Data{
				PublicKey:             bytesutil.SafeCopyBytes(deposit.PublicKey),
				WithdrawalCredentials: bytesutil.SafeCopyBytes(deposit.WithdrawalCredentials),
				Amount:                deposit.Amount,
				Signature:             bytesutil.SafeCopyBytes(deposit.Signature),
			})
			if err != nil {
				return errors.Wrap(err, "individual deposit signature verification failed")
			}
		}

		// Add validator to the registry if the signature is valid
		if validSignature {
			err = AddValidatorToRegistry(state, deposit.PublicKey, deposit.WithdrawalCredentials, deposit.Amount)
			if err != nil {
				return errors.Wrap(err, "failed to add validator to registry")
			}
		}
	}

	return nil
}
```

File: beacon-chain/core/blocks/deposit.go [hunk 7]
```diff
@@ -159,3 +173,44 @@ func verifyDepositDataWithDomain(ctx context.Context, deps []*ethpb.Deposit, dom
 	}
 	return nil
 }
+
+func verifyPendingDepositDataWithDomain(ctx context.Context, deps []*ethpb.PendingDeposit, domain []byte) error {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/blocks/deposit.go:177]: Added one for batch ProcessNewPendingDeposits, can add more

File: beacon-chain/core/electra/deposits.go [hunk 8]
```diff
@@ -82,7 +82,7 @@ func ProcessDeposits(
 //	  amount=deposit.data.amount,
 //	  signature=deposit.data.signature,
 //	 )
-func ProcessDeposit(beaconState state.BeaconState, deposit *ethpb.Deposit, verifySignature bool) (state.BeaconState, error) {
+func ProcessDeposit(beaconState state.BeaconState, deposit *ethpb.Deposit, verified bool) (state.BeaconState, error) {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/electra/deposits.go:85]: i creates circular dependencies if i put it in the blocks package

File: beacon-chain/core/blocks/deposit.go [hunk 9]
```diff
@@ -55,12 +55,26 @@ func BatchVerifyDepositsSignatures(ctx context.Context, deposits []*ethpb.Deposi
 		return false, err
 	}
 
-	verified := false
 	if err := verifyDepositDataWithDomain(ctx, deposits, domain); err != nil {
 		log.WithError(err).Debug("Failed to batch verify deposits signatures, will try individual verify")
-		verified = true
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/blocks/deposit.go:61]: https://github.com/prysmaticlabs/prysm/pull/14526 now as part of this pr

File: beacon-chain/core/blocks/deposit.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/blocks/deposit.go:177]: added more tests

File: config/params/loader_test.go [hunk 10]
```diff
@@ -180,6 +180,7 @@ func TestModifiedE2E(t *testing.T) {
 
 func TestLoadConfigFile(t *testing.T) {
 	t.Run("mainnet", func(t *testing.T) {
+		t.Skip("TODO: add back in after all spec test features are in.")
```
Comment (james-prysm) [inline config/params/loader_test.go:183]: not sure if https://github.com/prysmaticlabs/prysm/pull/14509 solves this one

**Review by terencechain — APPROVED** [review]

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**